pkg/controller/endpoint.go

```diff
@@ -235,14 +235,35 @@ func getServicePortBackends(endpoints *v1.Endpoints, servicePort v1.ServicePort,
 		}
 
 		for _, address := range subset.Addresses {
-			if util.CheckProtocol(serviceIP) == util.CheckProtocol(address.IP) {
-				if util.CheckProtocol(address.IP) == kubeovnv1.ProtocolIPv6 {
-					backends = append(backends, fmt.Sprintf("[%s]:%d", address.IP, targetPort))
-				} else {
-					backends = append(backends, fmt.Sprintf("%s:%d", address.IP, targetPort))
+			if address.TargetRef == nil || address.TargetRef.Kind != "Pod" {
+				backends = append(backends, util.JoinHostPort(address.IP, targetPort))
+				continue
+			}
+
+			var ip string
+			for _, pod := range pods {
+				if pod.Name == address.TargetRef.Name {
+					podIPs := pod.Status.PodIPs
+					if len(podIPs) == 0 && pod.Status.PodIP != "" {
+						podIPs = []v1.PodIP{{IP: pod.Status.PodIP}}
+					}
+					for _, podIP := range podIPs {
+						if util.CheckProtocol(podIP.IP) == protocol {
+							ip = podIP.IP
+							break
+						}
+					}
+					break
 				}
 			}
+			if ip == "" {
```

Comment:
why not using addresses.IP directly?

Reply:
In dual stack cluster, endpoints address contains IPv4 only.